geopmdpy/geopmdpy/stats.py

```diff
@@ -0,0 +1,152 @@
+#  Copyright (c) 2015 - 2024 Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+
+"""Wrapper for the geopm_stats_collector_*() C APIs
+
+"""
+
+import sys
+import os
+import errno
+import math
+from . import gffi
+from . import error
+
+gffi.gffi.cdef("""
+
+struct geopm_stats_collector_s;
+
+int geopm_stats_collector_create(size_t num_requests, const struct geopm_request_s *requests,
+                                 struct geopm_stats_collector_s **collector);
+
+int geopm_stats_collector_update(struct geopm_stats_collector_s *collector);
+
+int geopm_stats_collector_report_yaml(const struct geopm_stats_collector_s *collector,
+                                      size_t *max_report_size, char *report_yaml);
+
+int geopm_stats_collector_reset(struct geopm_stats_collector_s *collector);
+
+int geopm_stats_collector_free(struct geopm_stats_collector_s *collector);
+
+""")
+_dl = gffi.get_dl_geopmd()
+
+import json
+import yaml
+
+class Collector:
+    """ Object for aggregating statistics gathered from the PlatformIO interface of GEOPM
+
+    """
+    def __init__(self, signal_config):
+        """Create stats collector
+
+        Provide a list of read requests for PlatformIO.
+
+        Args:
+            signal_config (list((str, int, int))): List of requested
+                signals where each tuple represents
+                (signal_name, domain_type, domain_idx).
+
+        """
+        global _dl
+        self._collector_ptr = None
+        signal_config = [(rr[0], rr[1], rr[2]) for rr in signal_config]
+        num_signal = len(signal_config)
+        if num_signal == 0:
+            raise RuntimeError('Collector creation failed: length of input is zero')
+
+        signal_config_carr = gffi.gffi.new(f'struct geopm_request_s[{num_signal}]')
```

Comment:
```suggestion
        signal_config_carr = gffi.gffi.new('struct geopm_request_s[]', num_signal)
```
Not sure if this really matters. The cffi docs recommend it this way because it causes unnecessary re-parsing of their internally-cached first-arg strings. Since we'll likely only call this once per process that parser cache won't ever hit anyways.